fulltext/__init__.py

```diff
@@ -236,19 +241,67 @@ def backend_from_mime(mime):
     return mod
 
 
-def backend_from_ext(ext, ignore_err=True):
-    if ext and not ext.startswith('.'):
-        ext = "." + ext
+def backend_from_fname(name):
+    """Determine backend module object from a file name."""
+    ext = splitext(name)[1]
     try:
         mime = EXTS_TO_MIMETYPES[ext]
     except KeyError:
-        if not ignore_err:
-            raise ValueError("don't know how to handle %r extension" % ext)
-        warn("don't know how to handle %r extension; assume binary" % ext)
-        mime = DEFAULT_MIME
-    mod_name = MIMETYPE_TO_BACKENDS[mime]
-    mod = __import__(mod_name, fromlist=[' '])
-    return mod
+        with open(name, 'rb') as f:
+            return backend_from_fobj(f)
+    else:
+        mod_name = MIMETYPE_TO_BACKENDS[mime]
+        mod = __import__(mod_name, fromlist=[' '])
+        return mod
+
+
+def backend_from_fobj(f):
+    """Determine backend module object from a file object."""
+    if magic is None:
+        warn("magic lib is not installed; assuming mime type %r" % (
+            DEFAULT_MIME))
+        return backend_from_mime(DEFAULT_MIME)
+    else:
+        offset = f.tell()
+        try:
+            f.seek(0)
+            chunk = f.read(MAGIC_BUFFER_SIZE)
+            mime = magic.from_buffer(chunk, mime=True)
```

Comment:
`magic` has a `from_file()` function...

Reply:
In here we're dealing with a file object though

Reply:
You're right, my bad.